### PR TITLE
Update MSC3912 implementation to use `with_rel_type` instead of `with_relations`

### PR DESCRIPTION
**Pull request by weeman1337**
Adapts the implementation to the latest version of [MSC3912](https://github.com/matrix-org/matrix-spec-proposals/pull/3912#issuecomment-1567173949)

- `with_relations` → `with_rel_types`
- `with_rel_types` is now always a list of strings incl. `*`
- I decided to keep `with_relations` and mark it as deprecated. It should be removed once the MSC becomes stable.

Part of https://github.com/vector-im/element-web/issues/25471

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Update MSC3912 implementation to use `with_rel_type` instead of `with_relations` ([\#3420](https://github.com/matrix-org/matrix-js-sdk/pull/3420)).

### CODE REVIEW

**Comment by richvdh** [comment]
@weeman1337 a gentle request to try to make your PR subjects meaningful, and not to mark them as T-Task where there are functional changes :)

File: src/client.ts [hunk 1]
```diff
@@ -4627,22 +4627,24 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
         const reason = opts?.reason;
+        let withRelTypes = opts?.with_rel_types ?? opts?.with_relations;
 
-        if (
-            opts?.with_relations &&
-            this.canSupport.get(Feature.RelationBasedRedactions) === ServerSupport.Unsupported
-        ) {
+        if (typeof withRelTypes === "string") {
+            withRelTypes = [withRelTypes];
+        }
```
Comment (richvdh) [inline src/client.ts:4634]: why is this required? `IRedactOpts` does not permit a string here?

File: src/client.ts [hunk 2]
```diff
@@ -4627,22 +4627,24 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
+
+        if (withRelTypes && this.canSupport.get(Feature.RelationBasedRedactions) === ServerSupport.Unsupported) {
```
Comment (richvdh) [inline src/client.ts:4636]: implicit cast to bool is forbidden by our code style.

Likewise below

File: spec/unit/matrix-client.spec.ts [hunk 3]
```diff
@@ -1394,36 +1397,40 @@ describe("MatrixClient", function () {
...
-                    httpLookups = [
-                        {
-                            method: "PUT",
-                            path:
-                                `/rooms/${encodeURIComponent(roomId)}/redact/${encodeURIComponent(eventId)}` +
-                                `/${encodeURIComponent(txnId)}`,
-                            expectBody: {
-                                reason: "redaction test",
-                                [MSC3912_RELATION_BASED_REDACTIONS_PROP.unstable!]: [RelationType.Reference],
+                    it("should send with_rel_types in the request body", async () => {
```
Comment (richvdh) [inline spec/unit/matrix-client.spec.ts:1410]: ... except it doesn't, it uses the unstable prop name for now

File: spec/unit/matrix-client.spec.ts [hunk 4]
```diff
@@ -1394,36 +1397,40 @@ describe("MatrixClient", function () {
...
+
+                        httpLookups = [
+                            {
+                                method: "PUT",
+                                path:
+                                    `/rooms/${encodeURIComponent(roomId)}/redact/${encodeURIComponent(eventId)}` +
+                                    `/${encodeURIComponent(txnId)}`,
+                                expectBody: {
+                                    reason: "redaction test",
+                                    [MSC3912_RELATION_BASED_REDACTIONS_PROP.unstable!]: [RelationType.Reference],
```
Comment (richvdh) [inline spec/unit/matrix-client.spec.ts:1421]: to ensure that we aren't making the same mistake twice (and to make the test clearer), I think it is better to hardcode the property names here:

```suggestion
                                    "org.matrix.msc3912.with_relations": ["m.reference"],
```

Likewise the reference to `RelationType` above.

File: src/client.ts [hunk 5]
```diff
@@ -4627,22 +4627,24 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
                 "Server does not support relation based redactions " +
                     `roomId ${roomId} eventId ${eventId} txnId: ${txnId} threadId ${threadId}`,
             );
         }
 
-        const withRelations = opts?.with_relations
+        const withRelTypesContent = withRelTypes
             ? {
                   [this.canSupport.get(Feature.RelationBasedRedactions) === ServerSupport.Stable
                       ? MSC3912_RELATION_BASED_REDACTIONS_PROP.stable!
```
Comment (richvdh) [inline src/client.ts:4646]: my understanding is that this line is currently unreachable, given that `Feature.RelationBasedRedactions` has no defined `matrixVersion` - is that correct?

Not necessarily asking you to change anything here. Though I do think it would be clearer to factor out a `propName` variable instead of putting the ternary condition inline.

File: src/client.ts [hunk 6]
```diff
@@ -4627,22 +4627,24 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
-                      : MSC3912_RELATION_BASED_REDACTIONS_PROP.unstable!]: opts?.with_relations,
+                      : MSC3912_RELATION_BASED_REDACTIONS_PROP.unstable!]: withRelTypes,
               }
             : {};
```
Comment (richvdh) [inline src/client.ts:4649]: how about enclosing all this in an `if (!!withRelTypes)` block, to save repeatedly checking it?

**Review by richvdh — CHANGES_REQUESTED** [review]

File: src/client.ts (hunk 1 above)
Comment (weeman1337) [reply src/client.ts:4634]: The previous version allowed string or list. But since it is an unstable function anyway, I will remove the fallback.

File: src/client.ts (hunk 2 above)
Comment (weeman1337) [reply src/client.ts:4636]: Explicit cast to bool is forbidden by ESLint :clown_face: 

```
4633:13  error  Redundant double negation  no-extra-boolean-cast
```

File: src/client.ts (hunk 5 above)
Comment (weeman1337) [reply src/client.ts:4646]: > my understanding is that this line is currently unreachable, given that Feature.RelationBasedRedactions has no defined matrixVersion - is that correct?

Yes

> Not necessarily asking you to change anything here. Though I do think it would be clearer to factor out a propName variable instead of putting the ternary condition inline.

Done

File: src/client.ts (hunk 1 above)
Comment (richvdh) [reply src/client.ts]: > The previous version allowed string or list.

Are you sure? https://github.com/matrix-org/matrix-js-sdk/blob/e10db6f7e9978e2d8dd3640fbe2be503710fd506/src/%40types/requests.ts#L61 requires it to be an array.

File: src/client.ts (hunk 2 above)
Comment (richvdh) [reply src/client.ts]: Well, that's new. Surely we use double-negation everywhere?

I don't consider this resolved: we can't just ignore the style guide. Either we agree as a team to change it, or we respect it.

Anyway, in this case an explicit test for `undefined` is more appropriate.

File: src/client.ts [hunk 7]
```diff
@@ -4628,31 +4628,33 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
+            }
+
+            const withRelTypesPropName =
+                this.canSupport.get(Feature.RelationBasedRedactions) === ServerSupport.Stable
+                    ? MSC3912_RELATION_BASED_REDACTIONS_PROP.stable!
+                    : MSC3912_RELATION_BASED_REDACTIONS_PROP.unstable!;
+
+            withRelTypesContent = {
+                [withRelTypesPropName]: opts.with_rel_types,
+            };
```
Comment (richvdh) [inline src/client.ts:4648]:
```suggestion
            withRelTypesContent[withRelTypesPropName] =  opts.with_rel_types;
```

(and make `withRelTypesContent` a `const`)?

Indeed, consider having a `const content = { reason };` object and updating that here, instead of using `...` destructuring below. But that's gold-plating

File: src/client.ts [hunk 8]
```diff
@@ -4628,31 +4628,33 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
+        let withRelTypesContent = {};
 
-        const withRelations = opts?.with_relations
-            ? {
-                  [this.canSupport.get(Feature.RelationBasedRedactions) === ServerSupport.Stable
-                      ? MSC3912_RELATION_BASED_REDACTIONS_PROP.stable!
-                      : MSC3912_RELATION_BASED_REDACTIONS_PROP.unstable!]: opts?.with_relations,
-              }
-            : {};
+        if (opts?.with_rel_types) {
```
Comment (richvdh) [inline src/client.ts:4633]:
```suggestion
        if (opts?.with_rel_types !== undefined) {
```

File: spec/unit/matrix-client.spec.ts [hunk 9]
```diff
@@ -342,6 +339,11 @@ describe("MatrixClient", function () {
         store.getClientOptions = jest.fn().mockReturnValue(Promise.resolve(null));
         store.storeClientOptions = jest.fn().mockReturnValue(Promise.resolve(null));
         store.isNewlyCreated = jest.fn().mockReturnValue(Promise.resolve(true));
+
+        unstableFeatures = {
+            "org.matrix.msc3440.stable": true,
+        };
```
Comment (richvdh) [inline spec/unit/matrix-client.spec.ts:345]:
```suggestion
        unstableFeatures["org.matrix.msc3440.stable"] = true;
```

(though I'm not sure why it is being moved at all)

**Review by richvdh — APPROVED** [review]
some nit-picking but generally lgtm

File: src/client.ts (hunk 1 above)
Comment (weeman1337) [reply src/client.ts]: You are right. Also the MSC didn't allow it to be a string. Anyway removing this block then makes event more sense.

File: spec/unit/matrix-client.spec.ts (hunk 9 above)
Comment (weeman1337) [reply spec/unit/matrix-client.spec.ts:345]: This sets `unstableFeatures` to a defined state before each test. Another test sets an additional property on it: https://github.com/matrix-org/matrix-js-sdk/blob/deb908947b06047dc93bd5f82906e63133f9d3fd/spec/unit/matrix-client.spec.ts#L1400

Comment (weeman1337) [reply spec/unit/matrix-client.spec.ts:345]: → will add a comment

Comment (richvdh) [reply spec/unit/matrix-client.spec.ts:346]: oh, I see.

Comment (richvdh) [reply spec/unit/matrix-client.spec.ts:346]: oh, I see.
